kura/org.eclipse.kura.net.admin/src/main/java/org/eclipse/kura/net/admin/modem/telit/le910/TelitLe910.java

```diff
@@ -32,74 +34,160 @@
-            } catch (KuraException e) {
-                throw e;
-            } finally {
-                closeSerialPort(commAtConnection);
-            }
-        }
-        return simReady;
-    }
+	private static final Logger logger = LoggerFactory.getLogger(TelitLe910.class);
+
+	private boolean diversityEnabled;
+
+	public TelitLe910(ModemDevice device, String platform, ConnectionFactory connectionFactory) {
+		super(device, platform, connectionFactory);
+	}
+
+	@Override
+	public List<ModemTechnologyType> getTechnologyTypes() throws KuraException {
+
+		List<ModemTechnologyType> modemTechnologyTypes;
+		ModemDevice device = getModemDevice();
+		if (device == null) {
+			throw new KuraException(KuraErrorCode.UNAVAILABLE_DEVICE, "No modem device");
+		}
+		if (device instanceof UsbModemDevice) {
+			SupportedUsbModemInfo usbModemInfo = SupportedUsbModemsInfo.getModem((UsbModemDevice) device);
+			if (usbModemInfo != null) {
+				modemTechnologyTypes = usbModemInfo.getTechnologyTypes();
+			} else {
+				throw new KuraException(KuraErrorCode.UNAVAILABLE_DEVICE, "No usbModemInfo available");
+			}
+		} else {
+			throw new KuraException(KuraErrorCode.UNAVAILABLE_DEVICE, "Unsupported modem device");
+		}
+		return modemTechnologyTypes;
+	}
+
+	@Override
+	public boolean isSimCardReady() throws KuraException {
+
+		boolean simReady = false;
+		String port = null;
+
+		if (isGpsEnabled() && getAtPort().equals(getGpsPort()) && !getAtPort().equals(getDataPort())) {
+			port = getDataPort();
+		} else {
+			port = getAtPort();
+		}
+
+		synchronized (this.atLock) {
+			logger.debug("sendCommand getSimStatus :: {} command to port {}",
+					TelitHe910AtCommands.GET_SIM_STATUS.getCommand(), port);
+			byte[] reply = null;
+			CommConnection commAtConnection = null;
+			try {
+
+				commAtConnection = openSerialPort(port);
+				if (!isAtReachable(commAtConnection)) {
+					throw new KuraException(KuraErrorCode.NOT_CONNECTED, MODEM_NOT_AVAILABLE_FOR_AT_CMDS_MSG);
+				}
+
+				reply = commAtConnection.sendCommand(TelitHe910AtCommands.GET_SIM_STATUS.getCommand().getBytes(), 1000,
+						100);
+				if (reply != null) {
+					String simStatus = getResponseString(reply);
+					String[] simStatusSplit = simStatus.split(",");
+					if (simStatusSplit.length > 1 && Integer.valueOf(simStatusSplit[1]) > 0) {
+						simReady = true;
+					}
+				}
+			} catch (IOException e) {
+				throw new KuraException(KuraErrorCode.CONNECTION_FAILED, e);
+			} catch (KuraException e) {
+				throw e;
+			} finally {
+				closeSerialPort(commAtConnection);
+			}
+		}
+		return simReady;
+	}
+
+	@Override
+	public boolean isModemLTE() throws KuraException {
+		return true;
+	}
+
+	@Override
+	public boolean isDiversityEnabled() {
+		return this.diversityEnabled;
+	}
+
+	public void setDiversityEnabled(boolean diversityEnabled) {
+		this.diversityEnabled = diversityEnabled;
+	}
+
+	@Override
+	public void enableDiversity() throws KuraException {
```

Comment:
The enable and disable diversity code seem pretty much the same apart from minor things. Is it possible to optimise the code?